### ISSUE TITLE
Run functions right after thingShadow.register

### ISSUE THREAD

**Issue by Torsph** [aws/aws-iot-device-sdk-js#80]
Problem:
When starting the node application I register multiple shadows and delete the shadows to get a clean state. But when deleting the shadows I get random timeouts even if the Amazon IoT version of the shadow has been deleted (when I check the aws web console).

In the README.md I see a timeout and a comment:

```
// Note that the delay is not required for subsequent updates; only
// the first update after a Thing Shadow registration using default
// parameters requires a delay.  See API documentation for the update
// method for more details.
```

But I cannot find more information about the 5000ms delay and why it's there. When I add a delay of example 1000ms to the code below it never timeouts.

Could it be that the MQTT subscription to the /accepted topic isn't complete before AWS IoT publish a accepted message on the delete shadow and because of that the message is just lost? I have the same problem with get and update.

I tried to change the [thing code](https://github.com/aws/aws-iot-device-sdk-js/blob/master/thing/index.js) and make use of the callback in [_handleSubscriptions](https://github.com/aws/aws-iot-device-sdk-js/blob/master/thing/index.js#L134), but I got the same result.

Is there some documentation I have missed or is this an issue? If it's an issue how can it be fixed?

Another issue that I have seen is that I sometimes get `{"code":409,"message":"Version conflict"}` when updating the shadow after a delete. Guess it's the same as issue #68.

The code:

```
// Removed setup code

thingShadows.on('status',
    (thingName, stat, clientToken, stateObject) => {
        console.log(`[AWS - ${stat} - ${thingName}] -> ${JSON.stringify(stateObject)}`)
    })

thingShadows.on('timeout',
    (thingName, clientToken) => {
        console.log(`[AWS-timeout] -> timeout ${thingName} token ${clientToken}`)
    })

thingShadows.on('connect',
    () => {
        for (let thingNumber = 0; thingNumber < 3; thingNumber++) {
            let thingName = 'MyThingName' + thingNumber
            console.log(`[AWS-register] -> ${thingName}`)
            thingShadows.register(thingName)
            thingShadows.delete(thingName) // delete could be exchanged with update and get with same result   
        }
    })
```

Running the code:

```
// Add some data to the thingShadows
-> node index.js
[AWS-register] -> MyThingName0
[AWS-register] -> MyThingName1
[AWS-register] -> MyThingName2
[AWS - rejected - MyThingName0] -> {"code":404,"message":"No shadow exists with name: 'MyThingName0'"}
[AWS - accepted - MyThingName1] -> {"timestamp":1471417286}
[AWS] -> timeout MyThingName2 token Things2-2

// No data in the thingShadows
-> node index.js
[AWS-register] -> MyThingName0
[AWS-register] -> MyThingName1
[AWS-register] -> MyThingName2
[AWS - rejected - MyThingName0] -> {"code":404,"message":"No shadow exists with name: 'MyThingName0'"}
[AWS - rejected - MyThingName1] -> {"code":404,"message":"No shadow exists with name: 'MyThingName1'"}
[AWS - rejected - MyThingName2] -> {"code":404,"message":"No shadow exists with name: 'MyThingName2'"}

// Add some data to the thingShadows
-> node index.js
[AWS-register] -> MyThingName0
[AWS-register] -> MyThingName1
[AWS-register] -> MyThingName2
[AWS - rejected - MyThingName0] -> {"code":404,"message":"No shadow exists with name: 'MyThingName0'"}
[AWS] -> timeout MyThingName1 token Things2-1
[AWS] -> timeout MyThingName2 token Things2-2
```

**Comment by rongsaws** [comment]
> Could it be that the MQTT subscription to the /accepted topic isn't complete before AWS IoT publish a accepted message on the delete shadow

That's correct. The register function makes asynchronous calls to subscribe to the shadow topics, and it immediately returns without waiting for the acks from the gateway. If you then immediately calls delete/get/update, you will almost always get timeout. As you pointed out, the 5-second delay added to the sample in the README serves exactly the purpose of waiting for the acks to come back before starting shadow operations. 5 seconds is just an arbitrary delay; in reality it doesn't take that long for the acks to come back, that's why 1 second delay worked in your tests. 

If you want to use subscribe callback to determine when to start your first shadow operation, you are on the right track making changes to _handleSubscriptions(). One thing to note is you need to get acks for all the subscribe requests before moving onto the first shadow operation. If you have already done a patch for it, we are happy to review and incorporate it into the SDK.  

> Another issue that I have seen is that I sometimes get {"code":409,"message":"Version conflict"} when updating the shadow after a delete. Guess it's the same as issue #68.

Yes. We'll be releasing a fix for that soon.

**Comment by Torsph** [comment]
Thanks for the fast response. I will complete my first attempt on fixing the issue and send a PR :)
Hopefully within a day or two!

**Comment by Torsph** [comment]
Added PR #83 . Please review and provide feedback :)

**Comment by fengsongAWS** [comment]
Addressed in release v1.0.13.
